### ISSUE TITLE
Allow returning null value from IdResolver to make type information optional?

### ISSUE THREAD

**Issue by Antibrumm** [FasterXML/jackson-databind#633]
Current the TypeSerializerBase is throwing an exception in case the `IdResolver` does not generate a typeId. This makes it unnecessary difficult to implement a DefaultTyping where not all types need type information.

In my case only objects should have type information, arrays and scalars should not be touched. The `IdResolver` could handle this already, by checking the `value` in the `idForValue..` method, but the TypeSerializer will fail in case the resolver returns `null`. 

For this purpose i created an `OptionalAsPropertyTypeSerializer` that I activate via `ObjectMapper.setDefaultTyping`. The code simply checks for null values and reacts accordingly.

Quick example:
```java
class OptionalAsPropertyTypeSerializer extends TypeSerializerBase {

    @Override
    public void writeTypePrefixForObject(final Object value, final JsonGenerator jgen) throws IOException {
        final String typeId = idFromValue(value);
        if (typeId == null) {
            jgen.writeStartObject();
        } else if (jgen.canWriteTypeId()) {
            jgen.writeTypeId(typeId);
            jgen.writeStartObject();
        } else {
            jgen.writeStartObject();
            jgen.writeStringField(_typePropertyName, typeId);
        }
        setCurrentValueInContext(value, jgen);
    }
}
```

**Comment by Antibrumm** [comment]
I also took a look at this and from what i can tell from your implementation it will still write the result of the TypeIdResolver, just that it might be null.
Is this as you intended or should a `null` type id not be written at all? 

In the second case you could introduce the OptionalAsPropertyTypeSerializer as a new serializer, if you don't want to adapt all existing ones?

My current version is very specific in that case that it will only create type information for objects, never for scalars nor for arrays. I post the code anyway and I could work on it to take also the other types into account if you like.

    public class OptionalAsPropertyTypeSerializer extends TypeSerializerBase {
    
        /** The _type property name. */
        private final String _typePropertyName;
    
       public OptionalAsPropertyTypeSerializer(final TypeIdResolver idRes, final BeanProperty property,
            final String propName) {
            super(idRes, property);
            if (propName != null) {
                _typePropertyName = propName;
            } else {
                _typePropertyName = "_class";
            }
        }
    
        @Override
        public OptionalAsPropertyTypeSerializer forProperty(final BeanProperty prop) {
            return (_property == prop) ? this : new OptionalAsPropertyTypeSerializer(this._idResolver, prop,
                    this._typePropertyName);
        }
    
        @Override
        public String getPropertyName() {
            return _typePropertyName;
        }
    
        @Override
        public As getTypeInclusion() {
            return As.PROPERTY;
        }
    
        @Override
        protected String idFromValue(final Object value) {
            return _idResolver.idFromValue(value);
        }
    
        @Override
        protected String idFromValueAndType(final Object value, final Class<?> type) {
            return _idResolver.idFromValueAndType(value, type);
        }
    
        @Override
        public void writeCustomTypePrefixForArray(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
            jgen.writeStartArray();
        }
    
        @Override
        public void writeCustomTypePrefixForObject(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
            jgen.writeStartObject();
        }
    
        @Override
        public void writeCustomTypePrefixForScalar(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
        }
    
        @Override
        public void writeCustomTypeSuffixForArray(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
            if (!jgen.canWriteTypeId()) {
                writeTypeSuffixForArray(value, jgen); // standard impl works fine
            }
        }
    
        @Override
        public void writeCustomTypeSuffixForObject(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
            if (!jgen.canWriteTypeId()) {
                writeTypeSuffixForObject(value, jgen); // standard impl works fine
            }
        }
    
        @Override
        public void writeCustomTypeSuffixForScalar(final Object value, final JsonGenerator jgen, final String typeId)
                throws IOException {
            if (!jgen.canWriteTypeId()) {
                writeTypeSuffixForScalar(value, jgen); // standard impl works fine
            }
        }
    
        @Override
        public void writeTypePrefixForArray(final Object value, final JsonGenerator jgen) throws IOException {
            jgen.writeStartArray();
        }
    
        @Override
        public void writeTypePrefixForArray(final Object value, final JsonGenerator jgen, final Class<?> type)
                throws IOException {
            jgen.writeStartArray();
        }
    
        @Override
        public void writeTypePrefixForObject(final Object value, final JsonGenerator jgen) throws IOException {
            final String typeId = idFromValue(value);
            if (typeId == null) {
                jgen.writeStartObject();
            } else if (jgen.canWriteTypeId()) {
                jgen.writeTypeId(typeId);
                jgen.writeStartObject();
            } else {
                jgen.writeStartObject();
                jgen.writeStringField(_typePropertyName, typeId);
            }
        }
    
        @Override
        public void writeTypePrefixForObject(final Object value, final JsonGenerator jgen, final Class<?> type)
                throws IOException {
            final String typeId = idFromValueAndType(value, type);
            if (typeId == null) {
                jgen.writeStartObject();
            } else if (jgen.canWriteTypeId()) {
                jgen.writeTypeId(typeId);
                jgen.writeStartObject();
            } else {
                jgen.writeStartObject();
                jgen.writeStringField(_typePropertyName, typeId);
            }
        }
    
        @Override
        public void writeTypePrefixForScalar(final Object value, final JsonGenerator jgen) throws IOException {
        }
    
        @Override
        public void writeTypePrefixForScalar(final Object value, final JsonGenerator jgen, final Class<?> type)
                throws IOException {
        }
    
        @Override
        public void writeTypeSuffixForArray(final Object value, final JsonGenerator jgen) throws IOException {
            jgen.writeEndArray();
        }
    
        @Override
        public void writeTypeSuffixForObject(final Object value, final JsonGenerator jgen) throws IOException {
            jgen.writeEndObject();
        }
    
        @Override
        public void writeTypeSuffixForScalar(final Object value, final JsonGenerator jgen) throws IOException {
    
        }
    }

**Comment by cowtowncoder** [comment]
I see what you mean, and yes, I think you are right. I was not thinking through. I don't think writing of `null` type id makes sense.

**Comment by cowtowncoder** [comment]
I went ahead and retrofitted existing type serializers, since I think it does make sense to omit information if possible. Adding new type id serializers is possible if there is need; thank you for sharing the code.

**Comment by cowtowncoder** [comment]
Fix will be in 2.5.1
